File: src/registry.cpp
#include "openai_harmony/registry.hpp"
#include "openai_harmony/encoding.hpp"
#include "openai_harmony/tiktoken_ext.hpp"
#include <stdexcept>

namespace openai_harmony {

// Basic harmony encoding data - simplified version for demonstration
// In a real implementation, this would be loaded from external data files
std::unordered_map<std::vector<uint8_t>, Rank, VectorHash> create_basic_encoder() {
    std::unordered_map<std::vector<uint8_t>, Rank, VectorHash> encoder;
    
    // Add basic ASCII characters
    for (int i = 0; i < 256; ++i) {
        std::vector<uint8_t> byte = {static_cast<uint8_t>(i)};
        encoder[byte] = static_cast<Rank>(i);
    }
    
    // Add some common byte pairs (simplified BPE vocabulary)
    std::vector<std::pair<std::string, Rank>> common_pairs = {
        {"th", 256}, {"he", 257}, {"in", 258}, {"er", 259}, {"an", 260},
        {"re", 261}, {"ed", 262}, {"nd", 263}, {"on", 264}, {"en", 265},
        {"at", 266}, {"ou", 267}, {"it", 268}, {"is", 269}, {"or", 270},
        {"ti", 271}, {"as", 272}, {"te", 273}, {"et", 274}, {"ng", 275},
        {"of", 276}, {"al", 277}, {"de", 278}, {"se", 279}, {"le", 280},
        {"to", 281}, {"ar", 282}, {"st", 283}, {"nt", 284}, {"ro", 285},
        {"ne", 286}, {"om", 287}, {"li", 288}, {"la", 289}, {"el", 290},
        {"ma", 291}, {"ri", 292}, {"ic", 293}, {"co", 294}, {"ca", 295},
        {"tion", 296}, {"ing", 297}, {"and", 298}, {"the", 299}, {"for", 300},
        {"are", 301}, {"with", 302}, {"his", 303}, {"they", 304}, {"have", 305},
        {"this", 306}, {"will", 307}, {"you", 308}, {"that", 309}, {"but", 310},
        {"not", 311}, {"what", 312}, {"all", 313}, {"were", 314}, {"can", 315},
        {"had", 316}, {"her", 317}, {"was", 318}, {"one", 319}, {"our", 320}
    };
    
    for (const auto& [text, rank] : common_pairs) {
        std::vector<uint8_t> bytes(text.begin(), text.end());
        encoder[bytes] = rank;
    }
    
    return encoder;
}

std::unordered_map<std::string, Rank> create_harmony_special_tokens() {
    return {
        {"<|start|>", 200000},
        {"<|end|>", 200001},
        {"<|message|>", 200002},
        {"<|channel|>", 200003},
        {"<|constrain|>", 200004},
        {"<|call|>", 200005},
        {"<|reasoning|>", 200006},
        {"<|thinking|>", 200007},
        {"<|analysis|>", 200008},
        {"<|commentary|>", 200009},
        {"<|final|>", 200010},
        {"<|system|>", 200011},
        {"<|user|>", 200012},
        {"<|assistant|>", 200013},
        {"<|developer|>", 200014},
        {"<|tool|>", 200015}
    };
}

std::shared_ptr<HarmonyEncoding> load_harmony_encoding(HarmonyEncodingName name) {
    switch (name) {
        case HarmonyEncodingName::HarmonyGptOss: {
            // Create basic encoder and special tokens
            auto encoder = create_basic_encoder();
            auto special_tokens = create_harmony_special_tokens();
            
            // Create regex pattern for tokenization
            std::string pattern = R"([^\r\n\p{L}\p{N}]?+\p{L}++|\p{N}{1,3}| ?[^\s\p{L}\p{N}]++[\r\n]*|\s*[\r\n]|\s+(?!\S)|\s++)";
            
            // Create CoreBPE tokenizer
            auto tokenizer = std::make_shared<CoreBPE>(encoder, special_tokens, pattern);
            
            // Create format token mapping (empty for now)
            std::unordered_map<FormattingToken, std::string> format_token_mapping;
            
            // Create stop formatting tokens (empty for now)
            std::unordered_set<FormattingToken> stop_formatting_tokens;
            std::unordered_set<FormattingToken> stop_formatting_tokens_for_assistant_actions;
            
            // Create and return HarmonyEncoding with proper constructor
            return std::make_shared<HarmonyEncoding>(
                "harmony_gpt_oss",           // name
                8192,                        // n_ctx
                4096,                        // max_message_tokens
                1024,                        // max_action_length
                "o200k_harmony",             // tokenizer_name
                tokenizer,                   // tokenizer
                format_token_mapping,        // format_token_mapping
                stop_formatting_tokens,      // stop_formatting_tokens
                stop_formatting_tokens_for_assistant_actions  // stop_formatting_tokens_for_assistant_actions
            );
        }
        default:
            throw std::invalid_argument("Unknown encoding name");
    }
}

} // namespace openai_harmony

File: src/registry.rs
use std::{
    collections::{HashMap, HashSet},
    sync::Arc,
};

use crate::{
    encoding::{FormattingToken, HarmonyEncoding},
    tiktoken_ext,
};

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub enum HarmonyEncodingName {
    HarmonyGptOss,
}

impl std::fmt::Display for HarmonyEncodingName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                HarmonyEncodingName::HarmonyGptOss => "HarmonyGptOss",
            }
        )
    }
}

impl std::str::FromStr for HarmonyEncodingName {
    type Err = anyhow::Error;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "HarmonyGptOss" => Ok(HarmonyEncodingName::HarmonyGptOss),
            _ => anyhow::bail!("Invalid HarmonyEncodingName: {}", s),
        }
    }
}

impl std::fmt::Debug for HarmonyEncodingName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self}")
    }
}

#[cfg(all(not(target_arch = "wasm32"), feature = "network"))]
pub fn load_harmony_encoding(name: HarmonyEncodingName) -> anyhow::Result<HarmonyEncoding> {
    match name {
        HarmonyEncodingName::HarmonyGptOss => {
            let n_ctx = 1_048_576; // 2^20
            let max_action_length = 524_288; // 2^19
            let encoding_ext = tiktoken_ext::Encoding::O200kHarmony;
            Ok(HarmonyEncoding {
                name: name.to_string(),
                n_ctx,
                tokenizer: Arc::new(encoding_ext.load()?),
                tokenizer_name: encoding_ext.name().to_owned(),
                max_message_tokens: n_ctx - max_action_length,
                max_action_length,
                format_token_mapping: make_mapping([
                    (FormattingToken::Start, "<|start|>"),
                    (FormattingToken::Message, "<|message|>"),
                    (FormattingToken::EndMessage, "<|end|>"),
                    (FormattingToken::EndMessageDoneSampling, "<|return|>"),
                    (FormattingToken::Refusal, "<|refusal|>"),
                    (FormattingToken::ConstrainedFormat, "<|constrain|>"),
                    (FormattingToken::Channel, "<|channel|>"),
                    (FormattingToken::EndMessageAssistantToTool, "<|call|>"),
                    (FormattingToken::BeginUntrusted, "<|untrusted|>"),
                    (FormattingToken::EndUntrusted, "<|end_untrusted|>"),
                ]),
                stop_formatting_tokens: HashSet::from([
                    FormattingToken::EndMessageDoneSampling,
                    FormattingToken::EndMessageAssistantToTool,
                    FormattingToken::EndMessage,
                ]),
                stop_formatting_tokens_for_assistant_actions: HashSet::from([
                    FormattingToken::EndMessageDoneSampling,
                    FormattingToken::EndMessageAssistantToTool,
                ]),
            })
        }
    }
}

#[cfg(all(target_arch = "wasm32", feature = "network"))]
pub async fn load_harmony_encoding(name: HarmonyEncodingName) -> anyhow::Result<HarmonyEncoding> {
    match name {
        HarmonyEncodingName::HarmonyGptOss => {
            let n_ctx = 1_048_576; // 2^20
            let max_action_length = 524_288; // 2^19
            let encoding_ext = tiktoken_ext::Encoding::O200kHarmony;
            Ok(HarmonyEncoding {
                name: name.to_string(),
                n_ctx,
                tokenizer: Arc::new(encoding_ext.load().await?),
                tokenizer_name: encoding_ext.name().to_owned(),
                max_message_tokens: n_ctx - max_action_length,
                max_action_length,
                format_token_mapping: make_mapping([
                    (FormattingToken::Start, "<|start|>"),
                    (FormattingToken::Message, "<|message|>"),
                    (FormattingToken::EndMessage, "<|end|>"),
                    (FormattingToken::EndMessageDoneSampling, "<|return|>"),
                    (FormattingToken::Refusal, "<|refusal|>"),
                    (FormattingToken::ConstrainedFormat, "<|constrain|>"),
                    (FormattingToken::Channel, "<|channel|>"),
                    (FormattingToken::EndMessageAssistantToTool, "<|call|>"),
                    (FormattingToken::BeginUntrusted, "<|untrusted|>"),
                    (FormattingToken::EndUntrusted, "<|end_untrusted|>"),
                ]),
                stop_formatting_tokens: HashSet::from([
                    FormattingToken::EndMessageDoneSampling,
                    FormattingToken::EndMessageAssistantToTool,
                    FormattingToken::EndMessage,
                ]),
                stop_formatting_tokens_for_assistant_actions: HashSet::from([
                    FormattingToken::EndMessageDoneSampling,
                    FormattingToken::EndMessageAssistantToTool,
                ]),
            })
        }
    }
}

/// Load a [`HarmonyEncoding`] from raw tiktoken vocab bytes.
///
/// This is useful in environments where filesystem access and async HTTP
/// are unavailable (e.g. `wasm32-unknown-unknown` without JS bindings).
pub fn load_harmony_encoding_from_vocab_bytes(
    name: HarmonyEncodingName,
    vocab_bytes: &[u8],
) -> anyhow::Result<HarmonyEncoding> {
    match name {
        HarmonyEncodingName::HarmonyGptOss => {
            let n_ctx = 1_048_576;
            let max_action_length = 524_288;
            let encoding_ext = tiktoken_ext::Encoding::O200kHarmony;
            let mut specials: Vec<(String, u32)> = encoding_ext
                .special_tokens()
                .iter()
                .map(|(s, r)| ((*s).to_string(), *r))
                .collect();
            specials.extend((200014..=201088).map(|id| (format!("<|reserved_{id}|>"), id)));
            let tokenizer = tiktoken_ext::load_encoding_from_bytes(
                vocab_bytes,
                None,
                specials,
                &encoding_ext.pattern(),
            )?;
            Ok(HarmonyEncoding {
                name: name.to_string(),
                n_ctx,
                tokenizer: Arc::new(tokenizer),
                tokenizer_name: encoding_ext.name().to_owned(),
                max_message_tokens: n_ctx - max_action_length,
                max_action_length,
                format_token_mapping: make_mapping([
                    (FormattingToken::Start, "<|start|>"),
                    (FormattingToken::Message, "<|message|>"),
                    (FormattingToken::EndMessage, "<|end|>"),
                    (FormattingToken::EndMessageDoneSampling, "<|return|>"),
                    (FormattingToken::Refusal, "<|refusal|>"),
                    (FormattingToken::ConstrainedFormat, "<|constrain|>"),
                    (FormattingToken::Channel, "<|channel|>"),
                    (FormattingToken::EndMessageAssistantToTool, "<|call|>"),
                    (FormattingToken::BeginUntrusted, "<|untrusted|>"),
                    (FormattingToken::EndUntrusted, "<|end_untrusted|>"),
                ]),
                stop_formatting_tokens: HashSet::from([
                    FormattingToken::EndMessageDoneSampling,
                    FormattingToken::EndMessageAssistantToTool,
                    FormattingToken::EndMessage,
                ]),
                stop_formatting_tokens_for_assistant_actions: HashSet::from([
                    FormattingToken::EndMessageDoneSampling,
                    FormattingToken::EndMessageAssistantToTool,
                ]),
            })
        }
    }
}

fn make_mapping<I>(iter: I) -> HashMap<FormattingToken, String>
where
    I: IntoIterator<Item = (FormattingToken, &'static str)>,
{
    iter.into_iter().map(|(k, v)| (k, v.to_string())).collect()
}
